### ISSUE TITLE
A way to notify shapes going wrong in log_prob computation in ELBO?

### ISSUE THREAD

**Issue by lumip** [pyro-ppl/numpyro#1230]
We've recently run into a subtle (modelling) bug with broadcasting in the ELBO loss.
The offending part in our model boiled down to something like this:
```
import numpyro.distributions as dists

from numpyro import plate, sample
from numpyro.handlers import trace, seed

import numpy as np

N = 100
d = 7
logits = np.random.randn(N)
ys = np.random.binomial(1, .5, size=(N, 1))
 
def model(logits, ys):
    with plate("batch", len(ys)):
        sample("ys", dists.Bernoulli(logits), obs=ys)

# [some SVI inference code with Trace_ELBO here]
 ```
 
Note that while `logits` are of shape `(N,)`, our `ys` had shape `(N,1)` from our preprocessing script. As a consequence, the ELBO-internal call to `Bernoulli(logits).log_prob(ys)`  broadcasts over `ys` and returns a `(N, N)` matrix, over which the ELBO / `log_density` then sums. The program does not produce any error but our inference obviously fails.

Now, this is an oversight by us and not strictly a bug in numpyro (there could be scenarios where this behavior could be desired by a user, I guess (although I find it unlikely)). However, due the log_prob being abstracted away so nicely, it is very difficult to track this down when it is not desired, as in our case (we have been struggling to solve our poor model performance for a long time before we actually figured out that this is happening). I therefore wanted to bring this up and ideally have a discussion on whether there is a way to make this easier to spot if it happens.

One thought I had off the top of my head was to add an optional `expected_log_prob_shape` kwargs to `sample` where the user can specify what shape they would expect coming out of `log_prob` and the program failing if this is not the case. There might even be a way to do this without the explicit argument based on the distributions batch and even shapes maybe? I think another useful thing that could help a bit would be to actually check given `obs` for shape consistency with `sample_shape` (if both are provided to `sample`) as a sanity check for the user.

**Comment by martinjankowiak** [comment]
i'd suggest using the [format_shapes](http://num.pyro.ai/en/latest/utilities.html?highlight=%20format_shapes#numpyro.util.format_shapes) utility function. see [here](https://github.com/pyro-ppl/numpyro/blob/145ed8373fe268798b352b98350ffc596a51e682/test/test_util.py#L149) for example usage.

i guess one possibility would be to have an argument to e.g. `SVI` that prints this information to stdout automatically once upon initialization

**Comment by fehiepsi** [comment]
I think @maw501 already addresses part of the shape-mismatched issue in #1171. One way is to enhance the utility that @maw501 added and raise a warning whenever user code has missing plate statements like in the case
```
    with plate("batch", len(ys)):
        sample("ys", dists.Bernoulli(logits), obs=ys)
```
where `ys` has rank 2 but only one plate is declared.

**Comment by fehiepsi** [comment]
> have an argument to e.g. SVI that prints this information to stdout automatically once upon initialization

Good idea! I guess when a plate is missing, we can print out this message together with the warning. (When all plates are correctly declared, I guess `format_shapes` is not useful)
